fwutil/main.py

```diff
@@ -0,0 +1,296 @@
+
+
+def install_fw(ctx, fw_path):
+    component = ctx.obj[COMPONENT_CTX_KEY]
+    component_path = PATH_SEPARATOR.join(ctx.obj[COMPONENT_PATH_CTX_KEY])
+
+    status = False
+
+    try:
+        click.echo("Installing firmware:")
+        click.echo(TAB + fw_path)
+        log_helper.log_fw_install_start(component_path, fw_path)
+        status = component.install_firmware(fw_path)
+        log_helper.log_fw_install_end(component_path, fw_path, status)
+    except Exception as e:
+        log_helper.log_fw_install_end(component_path, fw_path, False, e)
+        cli_abort(ctx, str(e))
+
+    if not status:
+        log_helper.print_error("Firmware install failed")
+        ctx.exit(EXIT_FAILURE)
+
+
+def download_fw(ctx, url):
+    filename, headers = None, None
+
+    component_path = PATH_SEPARATOR.join(ctx.obj[COMPONENT_PATH_CTX_KEY])
+
+    try:
+        click.echo("Downloading firmware:")
+        log_helper.log_fw_download_start(component_path, str(url))
+        filename, headers = url.retrieve()
+        log_helper.log_fw_download_end(component_path, str(url), True)
+    except Exception as e:
+        log_helper.log_fw_download_end(component_path, str(url), False, e)
+        cli_abort(ctx, str(e))
+
+    return filename
+
+
+def validate_fw(ctx, param, value):
+    if value == HELP:
+        cli_show_help(ctx)
+
+    url = URL(value)
+
+    if not url.is_url():
+        path = click.Path(exists=True)
+        path.convert(value, param, ctx)
+    else:
+        ctx.obj[URL_CTX_KEY] = url
+
+    return value
+
+
+# 'fw' subcommand
+@component.command()
+@click.option('-y', '--yes', 'yes', is_flag=True, show_default=True, help="Assume \"yes\" as answer to all prompts and run non-interactively")
+@click.argument('fw_path', metavar='<fw_path>', callback=validate_fw)
+@click.pass_context
+def fw(ctx, yes, fw_path):
+    """Install firmware from local binary or URL"""
+    if not yes:
+        click.confirm("New firmware will be installed, continue?", abort=True)
+
+    url = None
+
+    if URL_CTX_KEY in ctx.obj:
+        url = ctx.obj[URL_CTX_KEY]
+        fw_path = download_fw(ctx, url)
+
+    try:
+        install_fw(ctx, fw_path)
+    finally:
+        if url is not None and os.path.exists(fw_path):
+            os.remove(fw_path)
+
+
+# 'update' subgroup
+@cli.command()
+@click.option('-y', '--yes', 'yes', is_flag=True, show_default=True, help="Assume \"yes\" as answer to all prompts and run non-interactively")
+@click.option('-f', '--force', 'force', is_flag=True, show_default=True, help="Install firmware regardless the current version")
+@click.option('-i', '--image', 'image', type=click.Choice(["current", "next"]), default="current", show_default=True, help="Update firmware using current/next image")
+@click.pass_context
+def update(ctx, yes, force, image):
+    """Update platform firmware"""
+    aborted = False
+
+    try:
+        squashfs = None
+
+        try:
+            cup = None
```

Comment:
C/C++ developer detected!
since cup is initialized anyway no need to assign None to it